druid/src/widget/textbox/editable_text.rs

```diff
@@ -0,0 +1,326 @@
+// Copyright 2019 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Traits for text editing and a basic String implementation.
+
+use crate::unicode_segmentation::GraphemeCursor;
+use std::ops::Range;
+
+/// An EditableText trait.
+pub trait EditableText: Sized {
+    type Cursor: EditableTextCursor<Self>;
+
+    /// Create a cursor with a copy of the text and a offset position.
+    fn cursor(&self, position: usize) -> Self::Cursor;
+
+    /// Replace range with new text.
+    fn edit(&mut self, iv: Range<usize>, new: impl Into<String>);
+
+    /// Get slice of text at range.
+    fn slice(&self, iv: Range<usize>) -> Option<&str>;
+
+    /// Get length of text.
+    fn len(&self) -> usize;
+
+    /// Get the next grapheme offset from the given offset, if it exists.
+    fn prev_grapheme_offset(&self, offset: usize) -> Option<usize>;
+
+    /// Get the previous grapheme offset from the given offset, if it exists.
+    fn next_grapheme_offset(&self, offset: usize) -> Option<usize>;
+
+    /// Get the next codepoint offset from the given offset, if it exists.
+    fn prev_codepoint_offset(&self, offset: usize) -> Option<usize>;
+
+    /// Get the previous codepoint offset from the given offset, if it exists.
+    fn next_codepoint_offset(&self, offset: usize) -> Option<usize>;
+}
+
+impl EditableText for String {
+    type Cursor = StringCursor;
+
+    //TODO: this can panic so maybe we should return a Result?
```

Comment:
this should not be able to panic. I think you're thinking of the case where the new string is larger than the provided interval? In this case you would split self at the start of the interval, append the new string, and then append the remaining old string on that.